### PR TITLE
[PX4] AP and FBW flash through bootloader fix

### LINKED ISSUE #2935: PX4 flash module has some issues with ChibiOS and bootloader firmware upload
Dear Community,

I was working with PX4 firmware upload  on **px4fmu_5.0_chibios** using px4_flash module and I encountered some issues:

1. PX4 flash module **(px4_flash.xml)** doesn't support ChibiOS USB driver and it can only be used with STM32 arch:
https://github.com/paparazzi/paparazzi/blob/0b1ce4679e7160c3debd7fbc648c0f10d67deed5/conf/modules/px4_flash.xml#L20
2. PX4 uploader script **(px_uploader.py)** seems to be old and gets stuck when communicating with some version-5 bootloaders.
3. **Programming FBW through AP** doesn't work here at all and I am not sure why.
4. ChibiOS USB serial driver **descriptor strings** are not the same as the ones in STM32 arch:
https://github.com/paparazzi/paparazzi/blob/0b1ce4679e7160c3debd7fbc648c0f10d67deed5/sw/airborne/arch/chibios/usb_ser_hw.c#L177-L198
compared with:
https://github.com/paparazzi/paparazzi/blob/0b1ce4679e7160c3debd7fbc648c0f10d67deed5/sw/airborne/arch/stm32/usb_ser_hw.c#L215-L220
This causes some upload issues when using  **set_target.py** script to upload through PX4 bootloader.

### PR DESCRIPTION
**Pull request by celarco**
This PR fixes issues mentioned in #2935.

It was tested OK with Cuav-Pixhawk-5 (px4fmu_5.0_chibios).

fix #2935

### CODE REVIEW

**Comment by gautierhattenberger** [comment]
It looks fine at a quick glance, but I don't have any board to test this. Can someone try that on PX4 boards, v5 and older versions ?

**Comment by celarco** [comment]
> It looks fine at a quick glance, but I don't have any board to test this. Can someone try that on PX4 boards, v5 and older versions ?

I was able to find an old Pixhawk fmu-v2 board and that was OK too. @fvantienen could you test this on your Pixhawk boards?

File: sw/airborne/modules/px4_flash/px4_flash.c [hunk 1]
```diff
@@ -130,8 +134,11 @@ void px4flash_event(void)
     unsigned char target = FLASH_PORT->get_byte(FLASH_PORT->periph);
     if (target == '1') { //target ap
       //the target is the ap, so reboot to PX4 bootloader
+#if defined(CHIBIOS_MCU_ARCH_H)
+      NVIC_SystemReset();
```
Comment (fvantienen) [inline sw/airborne/modules/px4_flash/px4_flash.c:138]: There is now an easier way to do this:
```mcu_reboot(MCU_REBOOT_BOOTLOADER)```

For the stm32 arch this needs to be added still, but is relatively easy to add to `mcu_arch`.

**Review by fvantienen — COMMENTED** [review]
Will test this on a pixhawk 4

**Review by fvantienen — DISMISSED** [review]

**Comment by OpenUAS** [comment]
Thx

**Comment by celarco** [comment]
Thank you @gautierhattenberger  @fvantienen @OpenUAS  all.

File: sw/airborne/modules/px4_flash/px4_flash.c (hunk 1 above)
Comment (celarco) [reply sw/airborne/modules/px4_flash/px4_flash.c:138]: Ok, I'll fix that in a new commit. I'll also look into stm32 arch and try to rewrite that, the way you mentioned.

**Comment by OpenUAS** [comment]
@fvantienen any remaining comment or else merge it into master, that would be great. Can re-test on Pixhawk1 (FMU2.4) Pixhawk4(FMU5) when needed. Chibios as well as bare metal.

**Comment by celarco** [comment]
> @fvantienen any remaining comment or else merge it into master, that would be great. Can re-test on Pixhawk1 (FMU2.4) Pixhawk4(FMU5) when needed. Chibios as well as bare metal.

Unfortunately I didn't have enough time to write the needed functions for mcu_reboot functionality in STM32 arch. I suggest we merge these into master and leave that for later.  The original code for STM32 arch worked this way, so it is fine I guess. @fvantienen Is it ok with you?

**Comment by OpenUAS** [comment]
Merging this PR would be helpful as it is.

**Review by gautierhattenberger — APPROVED** [review]

**Pull request merged**